File: 프로그래머스/2/42586. 기능개발/기능개발.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds) {

    // 큐에 소요되는 일수 저장
    queue<int> q;
    for(int i = 0 ; i < progresses.size() ; i++) {
        int remainProgress = 100 - progresses[i];
        int days = remainProgress % speeds[i] == 0 ? remainProgress / speeds[i] : remainProgress / speeds[i] + 1;
        q.push(days);
    }
    
    vector<int> result;
    // 큐 순회
    while(!q.empty()) {
        int cnt = 1;
        int curDays = q.front();
        q.pop();
        
        while(!q.empty() && q.front() <= curDays) {
            // 다음 작업 처리
            cnt++;
            q.pop();
        }
        
        result.push_back(cnt);
    }
    return result;
}

File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.*;
class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        
        int len = progresses.length;
        Queue<Integer> q = new ArrayDeque<>();
        List<Integer> lis = new ArrayList<>();
        
        
        for(int i = 0 ; i < len ; i++){
            int days = (100 - progresses[i] + speeds[i] - 1) / speeds[i];
            q.add(days);
        }
        
        
        while(!q.isEmpty()){
            int cur = q.poll();
            int cnt = 1;
            // 빨리 끝날 수 있는거 전부
            while(!q.isEmpty() && q.peek() <= cur){
                q.poll();
                cnt++;
            }
            lis.add(cnt);
        }
        
        
        int[] answer = new int[lis.size()];
        for(int i = 0 ; i < answer.length ; i++){
            answer[i] = lis.get(i);
        }
        
        return answer;
    }
}